src/rgw/rgw_sal_motr.cc

```diff
@@ -2737,18 +2731,47 @@ int MotrAtomicWriter::complete(size_t accounted_size, const std::string& etag,
   string tenant_bkt_name = get_bucket_name(obj.get_bucket()->get_tenant(), obj.get_bucket()->get_name());
   // Insert an entry into bucket index.
   string bucket_index_iname = "motr.rgw.bucket.index." + tenant_bkt_name;
+
+  if( !info.versioning_enabled() )
+  {
+	std::string obj_type = "simple object";
+        MotrObject::Meta old_meta;
+  	rgw_bucket_dir_entry old_ent;
+  	bufferlist old_check_bl;	
+	// Before updating bucket index with entry for new object, check if
+  	// old object exists. Perform M0_IC_GET operation on bucket index.
+  	rc = store->do_idx_op_by_name(bucket_index_iname, M0_IC_GET,
+                                obj.get_name(), old_check_bl);
+  	if (rc == 0 && old_check_bl.length() > 0) {
+    		auto ent_iter = old_check_bl.cbegin();
+    		old_ent.decode(ent_iter);
+    		rgw::sal::Attrs dummy;
+    		decode(dummy, ent_iter);
+    		old_meta.decode(ent_iter);
+    		std::unique_ptr<rgw::sal::Object> old_obj = obj.get_bucket()->get_object(rgw_obj_key(obj.get_name()));
+    		rgw::sal::MotrObject *old_mobj = static_cast<rgw::sal::MotrObject *>(old_obj.get());
+    		if (old_ent.meta.category == RGWObjCategory::MultiMeta) {
+      			obj_type = "multipart object";
+      			old_mobj->set_category(RGWObjCategory::MultiMeta);
+    		}
+    	ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " exists" << dendl;
+    	// Delete old object
+    	rc = old_mobj->delete_object(dpp, NULL, y, true);
+	if (rc == 0) {
+      		ldpp_dout(dpp, 20) <<__func__<< ": Old " << obj_type << " ["
+          	<< old_mobj->get_name() <<  "] deleted succesfully" << dendl;
+    	} else {
+      		ldpp_dout(dpp, 0) << __func__<<": Failed to delete old " << obj_type << " ["
+          	<< old_mobj->get_name() <<  "]. Error = " << rc << dendl;
+      		return rc;
+    	}
+  	}
+  }
   rc = store->do_idx_op_by_name(bucket_index_iname,
                                 M0_IC_PUT, obj.get_key().to_str(), bl);
   if (rc == 0)
     store->get_obj_meta_cache()->put(dpp, obj.get_key().to_str(), bl);
 
-  if (rc == 0 && !info.versioning_enabled())
-      // Delete old object data if exists.
-      old_obj.delete_mobj(dpp);
-
-  if (rc != 0)
```

Comment:
Why this if block removed ? If we are failing to add to bucket index, we shd delete the object .. correct ?

Reply:
This is the old code, we are deleting the object first now and after that we are putting the data in cache, and if we not able to delete the object then we are returning from there itself,so no need of this code now.